### PR TITLE
Recipe list route

### PR DESCRIPTION
**Pull request by chloee122**
work on:
- Recipe list route
- Message for recipe form submission

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -23,7 +38,9 @@ function App() {
       </div>
       <div className="flex">
         <RecipeList recipes={recipes} />
-        <DisplayRecipe />
+        <Route path={recipePath}>
```
Comment (duyle95) [inline src/App.js]: I think this whole `recipePath` computation is unnecessarily convoluted 🤔 

Instead, we can do something like:
```
<Route path="/recipes">
```
and in `Route` component, you check whether the `currentPath` matches the [regex](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_expressions) made of `path`. This approach would look more natural and simpler.

File: src/App.js [hunk 2]
```diff
@@ -1,19 +1,34 @@
...
+    if (currentPath.includes("/recipes"))
+      setRecipePath(window.location.pathname);
+  }, [currentPath]);
 
   const handleSearch = async (term) => {
     const results = await searchRecipes(term);
     setRecipes(results.recipes);
   };
 
+  const recipeId = window.location.pathname.includes("/recipes")
```
Comment (duyle95) [inline src/App.js]: [minor] This can stay inside `DisplayRecipe`

File: src/utils/convertIngredient.js [hunk 3]
```diff
@@ -1,6 +1,6 @@
 const convertIngredient = (text) => {
-  if (!text) return;
   const parts = text.split(",");
+  if (parts.length !==3) return null
```
Comment (duyle95) [inline src/utils/convertIngredient.js]: [nit] Please setup [prettier](https://prettier.io/) in this project, so you don't have to care about small aesthetic update to the code.

File: src/component/Route.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import { useContext } from "react";
+import { NavigationContext } from "../context/NavigationContext";
+
+function Route({ children, path }) {
+  const { currentPath } = useContext(NavigationContext);
+
+  if (currentPath === path) return children;
```
Comment (duyle95) [inline src/component/Route.js]: Here is what I would suggest using regex matching

File: src/component/RecipeListItem.js [hunk 5]
```diff
@@ -1,11 +1,16 @@
 import { useContext } from "react";
 import SelectedRecipeContext from "../context/SelectedRecipeContext";
+import { NavigationContext } from "../context/NavigationContext";
 
 function RecipeListItem({ recipe }) {
-  const { selectRecipe } = useContext(SelectedRecipeContext);
+  // const { selectRecipe } = useContext(SelectedRecipeContext);
```
Comment (duyle95) [inline src/component/RecipeListItem.js]: Remove comments.

Comment (duyle95) [inline src/component/RecipeListItem.js]: [nit] You can also look into eslint for React and pre-commit hook to enforce eslint on commit.

File: src/component/RecipeFormModal.js [hunk 6]
```diff
@@ -43,8 +43,11 @@ const HANDLE_INPUT = "handle_input";
 const HANDLE_INGREDIENT_INPUT = "handle_ingredient_input";
 
 function RecipeFormModal({ onClose }) {
+  const [isLoading, setIsLoading] = useState(false);
+  const [errorMessage, setErrorMessage] = useState(null);
+  const [isSuccessful, setIsSuccessful] = useState(null);
```
Comment (duyle95) [inline src/component/RecipeFormModal.js]: `isSuccessful` is not a good name for a non-boolean variable

File: src/component/RecipeFormModal.js [hunk 7]
```diff
@@ -55,21 +58,48 @@ function RecipeFormModal({ onClose }) {
...
+    }
   };
 
+  useEffect(() => {
+    if (isSuccessful) {
+      setTimeout(() => {
+        onClose();
+      }, 2000);
+    }
+  }, [isSuccessful]);
```
Comment (duyle95) [inline src/component/RecipeFormModal.js]: [Minor] Missing `onClose` as deps

File: src/component/RecipeFormModal.js [hunk 8]
```diff
@@ -55,21 +58,48 @@ function RecipeFormModal({ onClose }) {
...
+      setIsLoading(true);
+
+      const ingredients = state.ingredients
+        .filter((ingredient) => ingredient !== "")
+        .map(convertIngredient);
+
+      if (ingredients.some((ingredient) => ingredient === null))
+        throw Error("Input format was not correct! Please try again :)");
+
+      const recipe = { ...state, ingredients };
```
Comment (duyle95) [inline src/component/RecipeFormModal.js]: [minor] You can consider grouping these line into a function like `makeRecipe`

File: src/component/DisplayRecipe.js [hunk 9]
```diff
@@ -1,19 +1,21 @@
...
 
-function DisplayRecipe({ recipeId }) {
+function DisplayRecipe() {
   const [selectedRecipe, setSelectedRecipe] = useState(null);
 
   const [selectedServings, setSelectedServings] = useState(
     selectedRecipe?.servings || 0
   );
+  const { currentPath } = useContext(NavigationContext);
+  const recipeId = currentPath.substring(9);
```
Comment (duyle95) [inline src/component/DisplayRecipe.js]: There is a general practice that there shouldn't be any [magic number](https://en.wikipedia.org/wiki/Magic_number_(programming)#:~:text=In%20computer%20programming%2C%20a%20magic,see%20List%20of%20file%20signatures) in the code. Please assign the number `9` with a constant, which should say why 9 is being used.

File: src/component/RecipeFormModal.js [hunk 10]
```diff
@@ -45,7 +44,7 @@ const HANDLE_INGREDIENT_INPUT = "handle_ingredient_input";
 function RecipeFormModal({ onClose }) {
   const [isLoading, setIsLoading] = useState(false);
   const [errorMessage, setErrorMessage] = useState(null);
-  const [isSuccessful, setIsSuccessful] = useState(null);
+  const [successMessage, setSuccessMessage] = useState(null);
```
Comment (duyle95) [inline src/component/RecipeFormModal.js]: Let's have default state for all `...Message` states to be empty string

File: src/component/RecipeFormModal.js [hunk 11]
```diff
@@ -45,7 +44,7 @@ const HANDLE_INGREDIENT_INPUT = "handle_ingredient_input";
...
   const [state, dispatch] = useReducer(recipeFormReducer, initialFormValue);
```
Comment (duyle95) [inline src/component/RecipeFormModal.js]: I think you can name something else instead of `state`, maybe `formState` for better readability?

File: src/component/Route.js [hunk 12]
```diff
@@ -4,7 +4,10 @@ import { NavigationContext } from "../context/NavigationContext";
 function Route({ children, path }) {
   const { currentPath } = useContext(NavigationContext);
 
-  if (currentPath === path) return children;
+  if (path.test(currentPath)) {
```
Comment (duyle95) [inline src/component/Route.js]: I usually use String [match](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/match) instead, so something like `currentPath.match(path)`. Remember to parse `path` to RegExp first, if you want to take my approach suggested [here](https://github.com/chloee122/forkify-react/pull/5#discussion_r1503307891).

File: src/utils/makeRecipe.js [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+    const ingredients = state.ingredients
+      .filter((ingredient) => ingredient !== "")
+      .map(convertIngredient);
+
+    if (ingredients.some((ingredient) => ingredient === null))
+      throw Error("Input format was not correct! Please try again :)");
+
+    const recipe = { ...state, ingredients };
+    return recipe;
+  } catch (err) {
```
Comment (duyle95) [inline src/utils/makeRecipe.js]: You prolly don't need to try catch here, because `makeRecipe` is already called within a try catch in `handleSubmit` 🤔

File: src/App.js [hunk 14]
```diff
@@ -1,34 +1,22 @@
...
 import { searchRecipes } from "./api";
 import RecipeSearchForm from "./component/RecipeSearchForm";
 import RecipeList from "./component/RecipeList";
 import DisplayRecipe from "./component/DisplayRecipe";
 import Bookmark from "./component/Bookmark";
 import RecipeAddButton from "./component/RecipeAddButton";
 import Route from "./component/Route";
-import { NavigationContext } from "./context/NavigationContext";
+
+const RECIPE_PATH = /\/recipes\//;
```
Comment (duyle95) [inline src/App.js]: My initial idea is that this should be a string instead, and in `Route` component, this string will be parsed to a RegExp with `const re = new RegExp("/recipes/")`. Your approach is also fine, but for example if you gonna create new routes, you have to define more regex, which is not ideal imo 🤔

File: src/App.js [hunk 15]
```diff
@@ -5,13 +5,27 @@ import RecipeList from "./component/RecipeList";
...
+    try {
+      setRecipes([]);
+      setIsLoading(true);
+      const results = await searchRecipes(term);
+      if (results.recipes.length === 0)
+        throw Error("No recipe was found! Try other recipes");
+      setIsLoading(false);
+      setRecipes(results.recipes);
+    } catch (err) {
+      console.error(err);
```
Comment (duyle95) [inline src/App.js]: No console.log

File: src/component/RecipeList.js [hunk 16]
```diff
@@ -27,9 +28,13 @@ function RecipeList({ recipes }) {
     );
   });
 
+  console.log(isLoading);
```
Comment (duyle95) [inline src/component/RecipeList.js]: Same

File: src/component/RecipeList.js [hunk 17]
```diff
@@ -27,9 +28,13 @@ function RecipeList({ recipes }) {
...
+  console.log(recipes);
   return (
     <div>
-      <div>{renderedRecipes}</div>
+      <div>
```
Comment (duyle95) [inline src/component/RecipeList.js]: [nit] I think you don't need to wrap the loader with div?

File: src/utils/makeRecipe.js [hunk 18]
```diff
@@ -4,20 +4,16 @@ const convertIngredient = (text) => {
...
+  if (ingredients.some((ingredient) => ingredient === null))
+    throw Error("Input format was not correct! Please try again :)");
 
-    const recipe = { ...state, ingredients };
-    return recipe;
-  } catch (err) {
-    throw err;
-  }
+  const newRecipe = { ...recipe, ingredients };
+  return newRecipe;
```
Comment (duyle95) [inline src/utils/makeRecipe.js]: you can also do it like this ` return { ...recipe, ingredients }`

File: src/utils/makeRecipe.test.js [hunk 19]
```diff
@@ -1,6 +1,57 @@
...
+      quantity: 1,
+      unit: "kg",
+      description: "Rice",
+    };
+    expect(convertIngredient("1,kg,Rice")).toEqual(expectedIngredientObj);
+  });
+});
+
+describe("makeRecipe", () => {
+  describe("processes a list of ingredient strings that has empty values", () => {
```
Comment (duyle95) [inline src/utils/makeRecipe.test.js]: I think this test suite is too simple to divide them into describe under describe. I think something like this would be fine (for most of the cases):
```
describe("makeRecipe", () => {
  it("should return correct recipe when processes a list of valid ingredient strings", () => {...}
  it("should return correct recipe when processing a list of ingredient strings that has empty values", () => {...}
  it("should throw errors when processing a list of ingredient strings that has empty values", () => {...}
```

Overall the format for test case name should be some thing like `should ... when ...`. I personally then write tests, starting from happy path to edge cases and lastly error cases

File: src/component/Recipe.js [hunk 20]
```diff
@@ -1,12 +1,64 @@
...
+import RecipeDetails from "./RecipeDetails";
+import RecipeIngredients from "./RecipeIngredients";
+import RecipeDirection from "./RecipeDirection";
+import * as api from "../api";
+import { NavigationContext } from "../context/NavigationContext";
+
+
+const STARTING_INDEX_OF_RECIPE_ID = 9;
+
+function DisplayRecipe() {
```
Comment (duyle95) [inline src/component/Recipe.js]: Please name the functional component the same as filename.

File: src/component/RecipeFormModal.js [hunk 21]
```diff
@@ -94,6 +94,15 @@ function RecipeFormModal({ onClose }) {
     }
   }, [successMessage, onClose]);
 
+  useEffect(() => {
+    document.body.classList.add("overflow-hidden");
+
+    return () => {
+      document.body.classList.remove("overflow-hidden");
+      console.log("eyyo");
```
Comment (duyle95) [inline src/component/RecipeFormModal.js]: Eyyo 👋

File: src/component/RecipeList.js [hunk 22]
```diff
@@ -28,13 +28,17 @@ function RecipeList({ recipes, isLoading, errorMessage }) {
     );
   });
 
-  console.log(isLoading);
-  console.log(recipes);
+  const renderedError = errorMessage ? (
```
Comment (duyle95) [inline src/component/RecipeList.js]: [non-blocking] you can consider to create a reusable `ErrorMessage` component

File: src/component/Bookmark.js [hunk 23]
```diff
@@ -16,7 +16,7 @@ function Bookmark() {
       onMouseLeave={() => setShowBookmarks(false)}
     >
       <button>BOOKMARKS</button>
-      {showBookmarks && <BookmarkList />}
+      {showBookmarks && <BookmarkList setShowBookmarks={setShowBookmarks} />}
```
Comment (duyle95) [inline src/component/Bookmark.js:19]: Maybe rename this `setShowBookmarks` props to `closeBookmarks`
